src/osp/id_registry.h

```diff
@@ -0,0 +1,341 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2021 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include "hierarchical_bitset.h"
+
+#include <cassert>
+#include <limits>
+#include <utility>
+#include <vector>
+
+namespace osp
+{
+
+template<class TYPE_T>
+constexpr TYPE_T id_null() noexcept
+{
+    using underlying_t = typename std::underlying_type_t<TYPE_T>;
+    return TYPE_T(std::numeric_limits<underlying_t>::max());
+}
+
+//-----------------------------------------------------------------------------
+
+/**
+ * @brief Generates reusable sequential IDs
+ */
+template<typename ID_T, bool NO_AUTO_RESIZE = false>
+class IdRegistry
+{
+    using id_int_t = std::underlying_type_t<ID_T>;
```

Comment:
same advice about static_assert(std::is_enum_v<ID_T>)

Reply:
Alternatively, you can make a new underlying_type type_trait that doesn't barf if it's given an int.

template<typename TYPE_T>
using type_identity
{
    using type = TYPE_T;
}
template<typename TYPE_T>
using type_identity_t = typename type_identity<TYPE_T>::type;

template<typename TYPE_T, typename = void>
struct underlying_int_type;

template<typename TYPE_T>
struct underlying_int_type<TYPE_T, std::enable_if<std::is_enum_v<TYPE_T>> : std:underlying_type<TYPE_T>{};

template<typename TYPE_T>
struct underlying_int_type<TYPE_T, std::enable_if< ! std::is_enum_v<TYPE_T>> : type_identity<TYPE_T>{};

Reply:
type_identity is defined in C++20, but not in C++17, so you could if-def this based on the C++ version we're compiling for if you felt like it.

Reply:
I think I understand that this is SFINAE, but I'm still going "wtf" in the inside. It kind of feels overkill and compile time hungry in this case.

Overall, I see the problem as a kind of compile-time else-if chain:
```
if (type is integer) { return type }
else if (type is enum) { return underlying_t<type> }
else { static assert }
```

Reply:
Na, this is just partial template specialization. Compilers can do this very cheaply and it's a lot less expensive than using std::conditional

Reply:
Let me know if that doesn't actually compile btw, since i copied it while also making some translations, from another file.

Reply:
The `std::enable_if` just had to be `std::enable_if_t` and a syntax error, otherwise it works well:

```cpp
template<typename TYPE_T>
struct type_identity
{
    using type = TYPE_T;
};

template<typename TYPE_T>
using type_identity_t = typename type_identity<TYPE_T>::type;

template<typename TYPE_T, typename = void>
struct underlying_int_type;

template<typename TYPE_T>
struct underlying_int_type< TYPE_T, std::enable_if_t< std::is_enum_v<TYPE_T> > >
 : std::underlying_type<TYPE_T>
{ };

template<typename TYPE_T>
struct underlying_int_type< TYPE_T, std::enable_if_t< ! std::is_enum_v<TYPE_T> > >
 : type_identity<TYPE_T>
{ };

template<typename TYPE_T>
using underlying_int_type_t = typename underlying_int_type<TYPE_T>::type;

// Test
enum class Test : uint32_t {};
static_assert(std::is_same_v<underlying_int_type_t<int>, int>);
static_assert(std::is_same_v<underlying_int_type_t<Test>, uint32_t>);
```